src/Game/AITools.h

```diff
@@ -0,0 +1,95 @@
+
+//
+// Created by tim on 14.06.19.
+//
+
+#ifndef KI_AITOOLS_H
+#define KI_AITOOLS_H
+
+#include <memory>
+#include <optional>
+#include <deque>
+#include <list>
+#include <functional>
+#include <set>
+
+namespace aiTools{
+    template <typename T>
+    class SearchNode {
+    public:
+        SearchNode(T state, std::optional<std::shared_ptr<SearchNode<T>>> parent, int pathCost) :
+            state(state), parent(parent), pathCost(pathCost) {}
+
+        T state;
+        std::optional<std::shared_ptr<SearchNode<T>>> parent;
+        int pathCost = 0;
+        bool operator==(const SearchNode<T> &other) const;
+
+    };
+
+
+    template<typename T>
+    bool SearchNode<T>::operator==(const SearchNode<T> &other) const {
+        return this->state == other.state;
+    }
+
+
+    /**
+     * A*algorithm implemented as graph search
+     * @tparam T Type of the SearchNode used
+     * @param Start initial state
+     * @param Destination goal state
+     * @param expand Function used for expanding nodes
+     * @param f Function used for evaluating a node. A high value corresponds to low utility
+     * @return The goal node with information about its parent or nothing if problem is impossible
+     */
+    template <typename T>
+    auto aStarSearch(SearchNode<T> &start, SearchNode<T> &destination,
```

Comment:
Könnte man des nicht in die cpp Auslagern so wie wir das auch mit den template Klassen im Game Model hatten? Also über Explicit instantiation von dem Zeug in der cpp